### PR TITLE
nextflow: 22.10.6 -> 24.08.0-edge + remove buildFHSEnv + compile from source + add tests

### PR DESCRIPTION
**Pull request by rollf**
~~There is no need to use buildFHSEnv (anymore) on Linux. I removed it here because a) it makes the code cleaner, b) improves access (e.g. the `version` will be shown on https://search.nixos.org/packages?query=nextflow) and c) allows to add & run tests.~~

**Update 2024-09-06**

This PR is a complete rewrite of the `nextflow` derivation:
* Do not use `buildFSHEnv` anymore but rather patch the original source (so that `/bin/bash` is not referenced anymore)
* The above is only possible when compiling from source (instead of downloading a release), this is done using `gradle`
* Update nextflow version
* Use the latest Java version
* Add tests

This PR supersedes [a previous one](https://github.com/NixOS/nixpkgs/pull/286701) prepared by @edmundmiller. I have incorporated his `openjdk` change.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/interpreters/nextflow/default.nix
Comment (edmundmiller) [inline pkgs/development/interpreters/nextflow/default.nix:54]: Think we could get away with a `nextflow run hello` too?

Comment (rollf) [reply pkgs/development/interpreters/nextflow/default.nix:54]: Mh, I think this needs internet access, right? I tried to run this and this yields:

```bash
❯ nix build path:.#nextflow.tests.hello
error: builder for '/nix/store/k7ic1jfrcmyiaxckk5ziqgcjvic1ckaw-nextflow-run-hello.drv' failed with exit code 1;
       last 10 log lines:
       > sourcing setup hook '/nix/store/cickvswrvann041nqxb0rxilc46svw1n-prune-libtool-files.sh'
       > sourcing setup hook '/nix/store/xyff06pkhki3qy1ls77w10s0v79c9il0-reproducible-builds.sh'
       > sourcing setup hook '/nix/store/aazf105snicrlvyzzbdj85sx4179rpfp-set-source-date-epoch-to-latest.sh'
       > sourcing setup hook '/nix/store/gps9qrh99j7g02840wv5x78ykmz30byp-strip.sh'
       >
       >  N E X T F L O W   ~  version 24.04.4
       >
       > Pulling nextflow-io/hello ...
       > WARN: Cannot read project manifest -- Cause: api.github.com
       > Cannot find `nextflow-io/hello` -- Make sure exists a GitHub repository at this address `https://github.com/nextflow-io/hello`
       For full logs, run 'nix log /nix/store/k7ic1jfrcmyiaxckk5ziqgcjvic1ckaw-nextflow-run-hello.drv'.
```

```diff
-  passthru.tests.version = testers.testVersion {
-    package = nextflow;
-    command = "HOME=$TMPDIR nextflow -version";
+  passthru.tests = {
+    version = testers.testVersion {
+      package = nextflow;
+      command = "HOME=$TMPDIR nextflow -version";
+    };
+    hello = runCommand "nextflow-run-hello" {nativeBuildInputs = [nextflow];} ''
+      HOME=$TMPDIR
+      nextflow run hello
+      touch $out
+    '';
   };
```

**Comment by rollf** [comment]
Ah, and upon further testing, not using `buildFSHEnv` doesn't work :(

File: pkgs/development/interpreters/nextflow/default.nix
Comment (rollf) [inline pkgs/development/interpreters/nextflow/default.nix:44]: I'm unhappy with this but could not get nextflow to compile otherwise. I don't know whether this is an issue with `mitm-cache` (i.e. maybe it does not correctly provide the dependencies) or an issue with nextflow's gradle config (i.e. it does not correctly list [transitive?] deps).

I was able to compile nextflow without issues in a `nix-shell` env with just `javac` & `make` at hand. But using the `mitm-cache` approach (as necessary for a derivation which should build deterministically), some deps were simply not fetched. For example, when removing `api "org.ow2.asm:asm-util:9.7"` (from the patch), the compilation aborts with

```bash
...nextflow> > Task :plugins:nf-wave:processResources
nextflow> > Task :nf-commons:compileGroovy FAILED
nextflow> > Task :plugins:nf-azure:copyPluginLibs
nextflow> > Task :plugins:nf-amazon:copyPluginLibs
nextflow> > Task :plugins:nf-google:copyPluginLibs
nextflow> FAILURE: Build failed with an exception.
nextflow> * What went wrong:
nextflow> Execution failed for task ':nf-commons:compileGroovy'.
nextflow> > Could not resolve all files for configuration ':nf-commons:detachedConfiguration1'.
nextflow>    > Could not find org.ow2.asm:asm-util:9.7.
nextflow>      Searched in the following locations:
nextflow>        - https://repo.maven.apache.org/maven2/org/ow2/asm/asm-util/9.7/asm-util-9.7.pom
nextflow>        - https://repo.eclipse.org/content/groups/releases/org/ow2/asm/asm-util/9.7/asm-util-9.7.pom
nextflow>        - https://oss.sonatype.org/content/repositories/snapshots/org/ow2/asm/asm-util/9.7/asm-util-9.7.pom
nextflow>        - https://s3-eu-west-1.amazonaws.com/maven.seqera.io/releases/org/ow2/asm/asm-util/9.7/asm-util-9.7.pom
nextflow>        - https://s3-eu-west-1.amazonaws.com/maven.seqera.io/snapshots/org/ow2/asm/asm-util/9.7/asm-util-9.7.pom
nextflow>      Required by:
nextflow>          project :nf-commons > org.apache.groovy:groovy-console:4.0.22
nextflow> * Try:
nextflow> > Run with --stacktrace option to get the stack trace.
nextflow> > Run with --info or --debug option to get more log output.
nextflow> > Run with --scan to get full insights.
nextflow> > Get more help at https://help.gradle.org.
nextflow> Deprecated Gradle features were used in this build, making it incompatible with Gradle 9.0.
nextflow> You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.
nextflow> For more on this, please refer to https://docs.gradle.org/8.8/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.
nextflow> BUILD FAILED in 40s
nextflow> 19 actionable tasks: 19 executed
```

I tested with

```bash
nix build -L .#nextflow.mitmCache.updateScript && ./result && nix build -L  .#nextflow
```

The `mitmCache.updateScript` runs throught but building itself yields the above errors.

Here is the diff to reproduce the error:

```diff
diff --git a/pkgs/development/interpreters/nextflow/add-build-dependencies.patch b/pkgs/development/interpreters/nextflow/add-build-dependencies.patch
index ccc774245d8c..cfd4d3629c16 100644
--- a/pkgs/development/interpreters/nextflow/add-build-dependencies.patch
+++ b/pkgs/development/interpreters/nextflow/add-build-dependencies.patch
@@ -2,7 +2,7 @@ diff --git a/modules/nf-commons/build.gradle b/modules/nf-commons/build.gradle
 index 2a5c6ec70..a29305106 100644
 --- a/modules/nf-commons/build.gradle
 +++ b/modules/nf-commons/build.gradle
-@@ -26,6 +26,12 @@ sourceSets {
+@@ -26,6 +26,11 @@ sourceSets {

  dependencies {
      api "ch.qos.logback:logback-classic:1.4.14"
@@ -10,7 +10,6 @@ index 2a5c6ec70..a29305106 100644
 +    api "org.apache.groovy:groovy-datetime:4.0.22"
 +    api "org.apache.groovy:groovy-dateutil:4.0.22"
 +    api "org.apache.groovy:groovy-sql:4.0.22"
-+    api "org.ow2.asm:asm-util:9.7"
 +    api "org.abego.treelayout:org.abego.treelayout.core:1.0.3"
      api "org.apache.groovy:groovy:4.0.22"
      api "org.apache.groovy:groovy-nio:4.0.22"
```

I basically just copy & pasted the "missing" deps from the error messages and build the patch accordingly.

**Comment by rollf** [comment]
@edmundmiller @Etjean I went down the rabbit hole of compiling nextflow on my own. I feel that the result is promising. Please check out the current state. I haven't sqashed my work yet.

**Comment by rollf** [comment]
@chayleaf Would you be able to have a look into the gradle dependency problem that I have explained [above](https://github.com/NixOS/nixpkgs/pull/339197#discussion_r1746481702)? Maybe we could somehow improve the situation.

File: pkgs/development/interpreters/nextflow/default.nix
Comment (rollf) [reply pkgs/development/interpreters/nextflow/default.nix:44]: Thank you for looking into this! Your suggestions (`gradleUpdateTask`) works fine. It is much better than a not-so-future-proof patch.

File: pkgs/development/interpreters/nextflow/default.nix
Comment (rollf) [reply pkgs/development/interpreters/nextflow/default.nix:51]: I don't think this will work. The groovy file is not [a script](https://github.com/NixOS/nixpkgs/blob/e1781b7c74f9f8e9606a6f9e422c4b51c6c49540/pkgs/stdenv/generic/setup.sh#L497) (and it's not about the firste line in the file but rather a string constant that is problematic) but `patchShebang` [wants](https://github.com/NixOS/nixpkgs/blob/19d4f7dc485f74109bd66ef74231285ff797a823/pkgs/build-support/setup-hooks/patch-shebangs.sh#L51) one.

File: pkgs/development/interpreters/nextflow/default.nix
Comment (rollf) [reply pkgs/development/interpreters/nextflow/default.nix:35]: Yes, you're right. I removed both. I check the others and found that `wget` as well as ata least one of the others (I commented out accordingly) are needed. Than I stopped investigating. I assume now that the dependencies haven't changed.

File: pkgs/development/interpreters/nextflow/default.nix
Comment (chayleaf) [reply pkgs/development/interpreters/nextflow/default.nix:43]: that's why i said install phase, not patch phase

Comment (rollf) [reply pkgs/development/interpreters/nextflow/default.nix:43]: @chayleaf My, maybe we are misunderstanding each other. The line I have to patch is this [here](https://github.com/nextflow-io/nextflow/blob/6e866ae81ff3bf8a9729e9dbaa9dd89afcb81a4b/modules/nextflow/src/main/groovy/nextflow/executor/BashWrapperBuilder.groovy#L87). This is not a shebang and it is neither used as one (i.e. not copied into a template file). From how I understand `patchShebang` (the bash function) works, it will not patch that file, no matter when/how I invoke it. It is not designed to replace arbitrary occurrences of strings that resemble a shebang within a given file but only the first line in "scripts" (see my links above).

Comment (chayleaf) [reply pkgs/development/interpreters/nextflow/default.nix:43]: you can patch the resulting bash wrapper, not the groovy file

Comment (rollf) [reply pkgs/development/interpreters/nextflow/default.nix:43]: IMHO, there is no resulting file that could be patched. `nextflow` generates the shell script on the fly (I believe this happens [here](https://github.com/nextflow-io/nextflow/blob/6e866ae81ff3bf8a9729e9dbaa9dd89afcb81a4b/modules/nextflow/src/main/groovy/nextflow/executor/local/LocalTaskHandler.groovy#L134)). Note that nextflow is a workflow engine, it generates a task graph and executes the tasks. A task is eventually only a (generated) script. This script is then executed. The string constant ("/bin/bash") is also used at other locations ([ex](https://github.com/nextflow-io/nextflow/blob/master/modules/nextflow/src/main/groovy/nextflow/processor/TaskBean.groovy#L136)).

**Comment by rollf** [comment]
@nbraud @NickCao You two have previously merged `nf-test`-PRs. Do you mind looking into this one and maybe merge it? Thanks.

File: pkgs/development/interpreters/nextflow/default.nix
Comment (rollf) [reply pkgs/development/interpreters/nextflow/default.nix]: I have added a `hello` test by simply downloading the necessary pipeline upfront.

File: pkgs/development/interpreters/nextflow/default.nix
Comment (SFrijters) [reply pkgs/development/interpreters/nextflow/default.nix:73]: Sort of related: I've been cleaning up `error: makeWrapper/makeShellWrapper must be in nativeBuildInputs` in cross-compilation scenarios, and while `makeWrapper` is correctly placed in `nativeBuildInputs` here, using `lib.makeBinPath` on it causes the error to appear in builds of reverse dependencies (in this case `nf-test`). I think it would make sense to keep only `makeWrapper` inside the actual `nativeBuildInputs` attribute and move the rest (or at least the ones that are necessary) to be just passed directly to `lib.makeBinPath`. Currently still running a build without your PR, but with this change.

Comment (rollf) [reply pkgs/development/interpreters/nextflow/default.nix:73]: (See this [commit](https://github.com/NixOS/nixpkgs/pull/341532/commits/b2b275a4ca9711351879a5d6b2db9aebd9530366) for an implementation of the suggested approach)

Comment (rollf) [reply pkgs/development/interpreters/nextflow/default.nix]: @SFrijters I adapted the code as you suggested and sucessfully run `nix build -L path:.#nextflow.tests.hello`. I have updated this PR now.

**Comment by rollf** [comment]
I realize that something is still wrong with the "fixed" shebang. While execution works fine when running on the host, using `docker.enabled = true` yields issues (`bash` not found). I'll have to investigate this :(

File: pkgs/development/interpreters/nextflow/default.nix
Comment (doronbehar) [inline pkgs/development/interpreters/nextflow/default.nix:48]: Typo:

```suggestion
  # During the build, some additional dependencies are downloaded ("detached configuration"). We thus need to run a full build on instead of the default one.
```

File: pkgs/development/interpreters/nextflow/default.nix
Comment (doronbehar) [inline pkgs/development/interpreters/nextflow/default.nix:84]: Could you switch to `versionCheckHook`? https://nixos.org/manual/nixpkgs/unstable/#versioncheckhook

File: pkgs/development/interpreters/nextflow/default.nix
Comment (doronbehar) [reply pkgs/development/interpreters/nextflow/default.nix:19]: > This version is compatible with Java 21. The current (as of 2024-09-06) nextflow [release](https://github.com/nextflow-io/nextflow/releases/tag/v24.04.4) does not yet support java21, but java19. The latter is not in nixpkgs(-unstable) anymore.

Could you explain that in a comment?

**Comment by doronbehar** [comment]
Looks good overall.

**Comment by rollf** [comment]
@doronbehar Thank you for your review. I'll address you remarks.

**Comment by rollf** [comment]
@doronbehar I adresse your concerns. The `versionCheckHook` fails. I somehow need to provde the env var `HOME` or `NXF_HOME` (see [docs](https://www.nextflow.io/docs/latest/config.html), or here). This is not possible IMHO with the hook. I tested setting up preVersionHook but the `env` anyway looses all environment variables. `nextflow` does not have another way other than using an env var. I don't see a mean to use the hook right now. Let me know what you think.

File: pkgs/development/interpreters/nextflow/default.nix
Comment (doronbehar) [inline pkgs/development/interpreters/nextflow/default.nix:17]: What is the purpose of this comment?

File: pkgs/development/interpreters/nextflow/default.nix
Comment (doronbehar) [inline pkgs/development/interpreters/nextflow/default.nix:18]: The whole point of using `finalAttrs` is that you won't have to use a `rec`. See:

https://nix.dev/guides/best-practices#recursive-attribute-set-rec

```suggestion
stdenv.mkDerivation (finalAttrs: {
```

File: pkgs/development/interpreters/nextflow/default.nix
Comment (doronbehar) [inline pkgs/development/interpreters/nextflow/default.nix:89]: > I don't see a mean to use the hook right now. Let me know what you think.

Yes so it seems.. I wonder whether this is maybe case is not that extreme, and we should support adding arbitrary arguments to the `env` command. For now, I'll leave it as it is, and we can go with `passthru.tests` on this one.

**Review by doronbehar — COMMENTED** [review]
Looks good, is this a draft because of the `versionCheckHook` failing?

File: pkgs/development/interpreters/nextflow/default.nix
Comment (rollf) [reply pkgs/development/interpreters/nextflow/default.nix:17]: Sorry, debug left-over.

File: pkgs/development/interpreters/nextflow/default.nix
Comment (rollf) [reply pkgs/development/interpreters/nextflow/default.nix]: Yes, I also wondered whether this package justifies adapting `versionCheckHook`. This is beyond my exptertise.

I have added & used `testers.testVersion` again.

**Comment by rollf** [comment]
> Looks good, is this a draft because of the `versionCheckHook` failing?

Yes. It it builds for me (locally) and I also integrated your remarks. Thanks again! The PR is ready from my point of view. I'll keep it as a draft for some time (this week?) since I'd like to test it more (using my local machine). I've had issues before (with docker, see above) that I only realized when actually using the package. I'd like to make sure it works at least for my use case.

Here is the last [diff](https://github.com/NixOS/nixpkgs/compare/66af0620ac211527d64a03dbbf957b52008613a4..f71f792df4d74eff19479323965cdee0e5635d50), btw.

File: pkgs/development/interpreters/nextflow/default.nix
Comment (doronbehar) [inline pkgs/development/interpreters/nextflow/default.nix:87]: nit:

```suggestion
  passthru.tests.nixos = nixosTests.nextflow;
```

File: pkgs/development/interpreters/nextflow/default.nix
Comment (doronbehar) [inline pkgs/development/interpreters/nextflow/default.nix:92]: nit2:

```suggestion
    command = "env HOME=$TMPDIR nextflow -version";
```

File: nixos/tests/nextflow.nix
Comment (doronbehar) [inline nixos/tests/nextflow.nix:59]: I'm pretty sure you aware of that due to the draft state, but I'm just noting that it'd be nice to take a decision here.

**Review by doronbehar — COMMENTED** [review]
Oh and BTW, it'd be nice to split the commits a bit. I'd do it like this:

```
nextflow: build from source
nextflow: add a passthru.tests.version
tests/nextflow: init
```

**Comment by rollf** [comment]
> Oh and BTW, it'd be nice to split the commits a bit. I'd do it like this:

Done. I also addressed your other remarks.

**Review by doronbehar — APPROVED** [review]
I'll be ready to merge this when you are :)

**Comment by rollf** [comment]
@doronbehar This is fine from my point of view now. I don't understand the merge conflict I'm currently seeing.
image

I'm converting this to "ready for review". I think it could be merged (assuming the merge conflict is a GitHub bug?).

 @edmundmiller and  @Etjean (the maintainers) haven't said anything though.

**Comment by doronbehar** [comment]
@rollf the merge conflict is not a mistake. I was because someone did a change that changed `stdenv.isLinux -> stdenv.hostPlatform.isLinux`, and then Git wasn't sure whether to delete these now new lines or not. I wanted to check that locally, and while I was there I fixed the conflict and force pushed. Please remind me again tomorrow to merge this when CI is green.

**Comment by SFrijters** [comment]
@doronbehar Ping as requested :)

File: nixos/tests/nextflow.nix
Comment (doronbehar) [inline nixos/tests/nextflow.nix:7]: This hash is different for aarch64-linux apparently (quoting from ofborg): 

```
error: hash mismatch in fixed-output derivation '/nix/store/75i49g2lqizqy7wynln75imsw7dhznfx-docker-image-quay.io-nextflow-bash-latest.tar.drv':
         specified: sha256-hZOe4qkUHZ4jPGp2Rlc0zG9uq00MYJ7xQV1KS4RLOpg=
            got:    sha256-oUiM5XRHWbogXfjmSMDaKvG2ZQy0d4GziFWyLFWc0C4=
```

I wonder if we should simply do this:

```suggestion
      sha256 = {
        x86_64-linux = "161s9f24njjx87qrwq0c9nmnwvyc6iblcxka7hirw78lm7i9x4w5";
        aarch64-linux = "sha256-oUiM5XRHWbogXfjmSMDaKvG2ZQy0d4GziFWyLFWc0C4=";
    }.${stdenv.hostPlatform.system} or (throw "unsupported system ${stdenv.hostPlatform.system}");
```

Or disable the test on platforms not `x86_64-linux`. I tend to the do the later, as it would be hard to fix further issues with aarch64-linux we might encounter, and I assume that both of us are losing patience :).

**Review by doronbehar — COMMENTED** [review]
Sorry to hold this further.. Found this issue:

File: nixos/tests/nextflow.nix
Comment (SFrijters) [reply nixos/tests/nextflow.nix:7]: Currently openjdk doesn't build for me in cross, so I don't think we'll be getting around to running the tests (or the application itself) in cross any time soon anyway. I don't have a big stake in this but I would recommend indeed just disabling the tests for cross.

Comment (doronbehar) [reply nixos/tests/nextflow.nix:7]: > Currently openjdk doesn't build for me in cross, so I don't think we'll be getting around to running the tests (or the application itself) in cross any time soon anyway. I don't have a big stake in this but I would recommend indeed just disabling the tests for cross.

We are not talking about cross compilation from x86_64-linux to aarch64-linux, but rather native building of the NixOS test derivation on aarch64-linux. `openjdk` builds fine on aarch64-linux according to [`hydra-check`](https://github.com/nix-community/hydra-check).

Comment (rollf) [reply nixos/tests/nextflow.nix:7]: Oh, wow. Why could the hash be different in the first place? I don't have an opinion here so I'd follow @doronbehar suggestions to only support `x86_64-linux`. I would disable the test within the `nextflow`-derivation itself (i.e. `passthru.tests.default = if host platform ...`), right?

Comment (doronbehar) [reply nixos/tests/nextflow.nix:7]: > I would disable the test within the `nextflow`-derivation itself (i.e. `passthru.tests.default = if host platform ...`), right?

That would make ofborg avoid building that derivation when it looks up `nextflow.passthru.tests`, but still your commit `nixos/nextflow` will make at least Hydra try to build that test on aarch64-linux. You can use the `handleTestOn` function that is used in `all-tests.nix`.

Comment (rollf) [reply nixos/tests/nextflow.nix:7]: I pushed a new commit using `handleTestOn`.

**Comment by doronbehar** [comment]
I have some more bad news, the x86_64-darwin build of `nextflow` itself fails. I also verified it on a real Darwin machine:

```
> Task :packOne FAILED
Attempt to copy /private/tmp/nix-build-nextflow-24.08.0-edge.drv-0/source/modules/nextflow/build/libs/nextflow-24.08.0-edge-one.jar to /var/empty/.nextflow/framework/24.08.0-edge/nextflow-24.08.0-edge-one.jar using NIO Channels failed due to 'failed to create the parent directory for /var/empty/.nextflow/framework/24.08.0-edge/nextflow-24.08.0-edge-one.jar'.  Falling back to streams.
FAILURE: Build failed with an exception.
* Where:
Script '/private/tmp/nix-build-nextflow-24.08.0-edge.drv-0/source/packing.gradle' line: 97
* What went wrong:
Execution failed for task ':packOne'.
> Failed to copy /private/tmp/nix-build-nextflow-24.08.0-edge.drv-0/source/modules/nextflow/build/libs/nextflow-24.08.0-edge-one.jar to /var/empty/.nextflow/framework/24.08.0-edge/nextflow-24.08.0-edge-one.jar due to java.nio.file.NoSuchFileException /var/empty/.nextflow/framework/24.08.0-edge/nextflow-24.08.0-edge-one.jar
```

And on Hydra the `nextflow` package doesn't fail to build (that's because it is much easier to build an FHS environment). Meaning this will be apparent for Darwin users. I really want to help and merge this, but I have no idea what are the chances that this package is used by some Darwin users, that will be annoyed to find out a change has broken their package.

**Comment by rollf** [comment]
@doronbehar I checked this. I think the issues stems from [here](https://github.com/nextflow-io/nextflow/blob/7c4538159dc7635f5c9bf3f0318f5dbdd0b7642f/packing.gradle#L30) in combination with this [copy operation](https://github.com/nextflow-io/nextflow/blob/7c4538159dc7635f5c9bf3f0318f5dbdd0b7642f/packing.gradle#L97). I wonder why this is not a problem on Linux. Maybe the gradle build is different. I just pushed a fix but forgot to compile it locally first. I am doing this right now. At least, I can check on Linux that it still works. Fingers crossed then for the x86_64-darwin build.

**Comment by doronbehar** [comment]
Similar error:

```
> Task :plugins:nf-console:compileGroovy
> Task :plugins:nf-console:classes
> Task :plugins:nf-codecommit:compileGroovy
> Task :plugins:nf-codecommit:classes
> Task :plugins:nf-amazon:classes
> Task :plugins:nf-tower:compileGroovy
> Task :plugins:nf-tower:classes
> Task :plugins:nf-wave:compileGroovy
> Task :plugins:nf-wave:classes
> Task :plugins:nf-google:compileGroovy
> Task :plugins:nf-google:classes

> Task :plugins:nf-azure:compileGroovy
07:28:45.138 [/127.0.0.1:53237 to /127.0.0.1:53235 workers] DEBUG com.azure.core.implementation.ReflectionUtils -- Attempting to use java.lang.invoke package to handle reflection.
07:28:45.146 [/127.0.0.1:53237 to /127.0.0.1:53235 workers] DEBUG com.azure.core.implementation.ReflectionUtils -- Successfully used java.lang.invoke package to handle reflection.

> Task :plugins:nf-azure:classes
> Task :compile
> Task :nextflow:shadowJar

> Task :packDist
+ Nextflow package `ALL` copied to: /private/tmp/nix-build-nextflow-24.08.0-edge.drv-0/source/build/releases/nextflow-24.08.0-edge-dist


> Task :packOne FAILED
Attempt to copy /private/tmp/nix-build-nextflow-24.08.0-edge.drv-0/source/modules/nextflow/build/libs/nextflow-24.08.0-edge-one.jar to /var/empty/.nextflow/framework/24.08.0-edge/nextflow-24.08.0-edge-one.jar usi>

FAILURE: Build failed with an exception.

* Where:
Script '/private/tmp/nix-build-nextflow-24.08.0-edge.drv-0/source/packing.gradle' line: 97

* What went wrong:
Execution failed for task ':packOne'.
> Failed to copy /private/tmp/nix-build-nextflow-24.08.0-edge.drv-0/source/modules/nextflow/build/libs/nextflow-24.08.0-edge-one.jar to /var/empty/.nextflow/framework/24.08.0-edge/nextflow-24.08.0-edge-one.jar du>

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.
> Get more help at https://help.gradle.org.
```

**Comment by doronbehar** [comment]
The above was taken from The [community's Darwin builder](https://nix-community.org/community-builder/)

**Comment by rollf** [comment]
I update my fix commit. I guess setting `HOME` did not work so I changed the gradle args. Still builds locally. Let's see what the darwin builder is saying.

**Review by doronbehar — APPROVED** [review]
Well done :tada: I tested it on the darwin build box and it worked :).

**Comment by stianlagstad** [comment]
Thank you all very much for the work on this! :100: :raised_hands:

**Comment by rollf** [comment]
Cool! Thank you @doronbehar for your support. It did feel like a team effort :+1: .

File: nixos/tests/nextflow.nix
Comment (rollf) [reply nixos/tests/nextflow.nix:7]: I think the hash is different because the data on the docker image differs among architectures/systems. Maybe one could have both hashes, depending on the `$system`. That might be something for the future.

**Pull request merged**
